File: Protogen/loxim_protocol/cpp/protocol/packages_data/BindingPackage.cpp
#include <string.h>
#include <stdlib.h>
#include "BindingPackage.h"
#include "ptools/PackageBufferWriter.h"
#include "ptools/PackageBufferReader.h"

using namespace protocol;

BindingPackage::BindingPackage(\
		CharArray* a_bindingName,\
		int64_t a_type,\
		Package* a_value
		)
{
	bindingName=a_bindingName;
	type=a_type;
	value=a_value;
}

BindingPackage::BindingPackage()
{
	bindingName=NULL;
	type=0;
	value=NULL;
}

BindingPackage::~BindingPackage()
{
	if (bindingName) delete bindingName;
	if(value) delete value;
	value=NULL;
}

void BindingPackage::serializeW(PackageBufferWriter *writer)
{
	Package::serializeW(writer);
	writer->writeSstring(bindingName);
	writer->writeVaruint(type);
	value->serializeContent(writer);
}

bool BindingPackage::deserializeR(PackageBufferReader *reader)
{
	if (!Package::deserializeR(reader)) return false;
	bindingName=reader->readSstring();
	{bool b; type=reader->readVaruint(b);}
	value=DataPackagesFactory::createPackageContent(type,reader);
	return true;
}

bool BindingPackage::equals(Package *_package)
{
	if (!_package) return false;
	if (_package->getPackageType()!=this->getPackageType()) return false;
	if (!Package::equals(_package)) return false;
	BindingPackage* p=(BindingPackage*)_package;
	if (!(CharArray::equal(getBindingName(),p->getBindingName()))) return false;
	if (!(getType()==p->getType())) return false;
	if (!((!getValue() && ! p->getValue())||(getValue() && p->getValue() && (getValue()->equals(p->getValue()))))) return false;
	return true;
}

File: Protogen/loxim_protocol/java/src/main/java/pl/edu/mimuw/loxim/protocol/packages_data/BindingPackage.java
package pl.edu.mimuw.loxim.protocol.packages_data;

import java.util.EnumSet;
import pl.edu.mimuw.loxim.protogen.lang.java.template.ptools.Package;
import pl.edu.mimuw.loxim.protogen.lang.java.template.exception.ProtocolException;
import pl.edu.mimuw.loxim.protogen.lang.java.template.streams.PackageInputStreamReader;
import pl.edu.mimuw.loxim.protogen.lang.java.template.streams.PackageOutputStreamWriter;


import pl.edu.mimuw.loxim.protocol.enums.*;


public  class BindingPackage extends Package {
		public final static long ID=(long)130;
	
    	/**
**/
    	private String bindingName=null;
    	/**
**/
    	private Long type=new Long(0);
    	/**
**/
    	private Package value=null;
    	
	public BindingPackage()
	{
	}

		public BindingPackage(
		    	String a_bindingName,
	        	Long a_type,
	        	Package a_value    	)
	{
        	bindingName=a_bindingName;
        	type=a_type;
        	value=a_value;
    
	}
		
		@Override
	public long getPackageType() {
		return (long)ID;
	}
		
	@Override
	protected void deserializeW(PackageInputStreamReader reader)
			throws ProtocolException {
		super.deserializeW(reader);
		
				try{
	    			bindingName=reader.readSString();
					type=reader.readVaruint();
					value=pl.edu.mimuw.loxim.protocol.packages_data.DataPackagesFactory.getInstance().createPackage(type);
value.deserializeContent(reader);
				}catch(Exception e)
		{
			throw new ProtocolException(e);
		}
			}
	
	@Override
	protected void serializeW(PackageOutputStreamWriter writer)
			throws ProtocolException {
		super.serializeW(writer);
				try{
	    			writer.writeSString(bindingName);		
					writer.writeVaruint(type);		
					value.serializeContent(writer);		
				}catch(Exception e)
		{
			throw new ProtocolException(e);
		}	
			}
	
	@Override
	public boolean equals(Object obj) {
		if(!super.equals(obj))
			return false;
		BindingPackage p=(BindingPackage)obj;
	    if (!(((bindingName==p.getBindingName())||((bindingName!=null)&&(p.getBindingName()!=null)&&((bindingName).equals(p.getBindingName())))))) 
return false;
		if (!(((type==p.getType())||((type!=null)&&(p.getType()!=null)&&((type).equals(p.getType())))))) 
return false;
		if (!(((value==p.getValue())||((value!=null)&&(p.getValue()!=null)&&((value).equals(p.getValue())))))) 
return false;
				return true;
	}
	
//================= GETTERS AND SETTERS ===================

        	public String getBindingName(){return bindingName;};
		
    	public void setBindingName(String a_bindingName){ bindingName=a_bindingName;};
		
        	public Long getType(){return type;};
		
    	public void setType(Long a_type){ type=a_type;};
		
        	public Package getValue(){return value;};
		
    	public void setValue(Package a_value){ value=a_value;};
		
    
	 
}


